<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <title>Undr Rover by rli99</title>
  </head>

  <body>
    <div id="container">
      <div class="inner">

        <header>
          <h1>Undr Rover</h1>
          <h2>Unmapped primer directed ROVER</h2>
        </header>

        <section id="downloads" class="clearfix">
          <a href="https://github.com/rli99/undr_rover/zipball/master" id="download-zip" class="button"><span>Download .zip</span></a>
          <a href="https://github.com/rli99/undr_rover/tarball/master" id="download-tar-gz" class="button"><span>Download .tar.gz</span></a>
          <a href="https://github.com/rli99/undr_rover" id="view-on-github" class="button"><span>View on GitHub</span></a>
        </section>

        <hr>

        <section id="main_content">
          <h2>
<a id="a-simple-fast-method-of-calling-variants-directly-from-fastqs-without-mapping" class="anchor" href="#a-simple-fast-method-of-calling-variants-directly-from-fastqs-without-mapping" aria-hidden="true"><span class="octicon octicon-link"></span></a>A simple, fast method of calling variants directly from FASTQs without mapping</h2>

<hr>

<h3>
<a id="how-to-use-undr-rover-to-call-variants" class="anchor" href="#how-to-use-undr-rover-to-call-variants" aria-hidden="true"><span class="octicon octicon-link"></span></a>How to use Undr Rover to call variants</h3>

<h4>
<a id="1-download-the-undr-rover-software" class="anchor" href="#1-download-the-undr-rover-software" aria-hidden="true"><span class="octicon octicon-link"></span></a>1. Download the Undr Rover software.</h4>

<p>Download Undr Rover as a .zip from <a href="https://github.com/rli99/undr_rover/archive/master.zip">GitHub</a></p>

<h4>
<a id="2-ensure-the-relevant-dependencies-are-installed" class="anchor" href="#2-ensure-the-relevant-dependencies-are-installed" aria-hidden="true"><span class="octicon octicon-link"></span></a>2. Ensure the relevant dependencies are installed.</h4>

<p>Undr Rover requires the Biopython, PyVCF and Pyfaidx libraries. They can be installed with the following line: <code>pip install -r REQUIREMENTS.txt</code></p>

<h4>
<a id="3-acquire-the-genome-reference-file-in-fasta-format" class="anchor" href="#3-acquire-the-genome-reference-file-in-fasta-format" aria-hidden="true"><span class="octicon octicon-link"></span></a>3. Acquire the genome reference file in FASTA format.</h4>

<p>As no mapping is undertaken by Undr Rover, this file is required for variants to be called. It can be downloaded at ___? Note that you will also need the .fai index file. </p>

<h4>
<a id="4-create-a-file-specifying-target-regions" class="anchor" href="#4-create-a-file-specifying-target-regions" aria-hidden="true"><span class="octicon octicon-link"></span></a>4. Create a file specifying target regions.</h4>

<p>The file should be tab-separated, and contain information in 5 columns. The first column will contain chromosome numbers, while the second and third columns specify the starting and ending coordinates of the target regions. The fourth and fifth columns will name the forward and reverse primers of each target region. </p>

<p>An example of such a file is shown below:</p>

<pre><code>chr11   31811401    31811498    PAX6_1_F1   PAX6_1_R1
chr11   31811499    31811597    PAX6_1_F2   PAX6_1_R2
chr11   31812210    31812307    PAX6_2_F1   PAX6_2_R1
chr11   31812308    31812406    PAX6_2_F2   PAX6_2_R2
chr11   31812399    31812496    PAX6_2_F3   PAX6_2_R3
chr11   31814945    31815042    PAX6_3_F1   PAX6_3_R1
chr11   31815043    31815140    PAX6_3_F2   PAX6_3_R2
chr11   31815147    31815244    PAX6_4_F1   PAX6_4_R1
chr11   31815242    31815339    PAX6_4_F2   PAX6_4_R2
chr11   31815340    31815437    PAX6_4_F3   PAX6_4_R3
</code></pre>

<p>Specify the name of this file using the --primer_coords command line parameter.</p>

<h4>
<a id="5-create-a-file-specifying-the-primer-sequences" class="anchor" href="#5-create-a-file-specifying-the-primer-sequences" aria-hidden="true"><span class="octicon octicon-link"></span></a>5. Create a file specifying the primer sequences.</h4>

<p>Undr Rover requires the primer sequences for each primer in order to correctly match reads to target regions. The file should be tab-separated and simply contain 2 columns, with the first containing the primer name and the second containing the primer sequence. </p>

<p>An example of such a file is shown below:</p>

<pre><code>PAX6_1_F1   TAACAGCCATTTTTCTTTCTTTCCT       
PAX6_1_R1   GTGAACCTGATATGTCTCAATACT        
PAX6_1_F2   TTTTTTTTTTTACTGTAATCTTGGCC      
PAX6_1_R2   TGTCTGTTTCTCAAAGGGAATATC        
PAX6_2_F1   GAGATTCTGTTTGGGTAAACTTCT        
PAX6_2_R1   CTACACCCCCCCACATATGCA       
PAX6_2_F2   TGGCTGACTGTTCATGTGTGTC      
PAX6_2_R2   TTTGCCTCTCTCCTCACAGCC       
PAX6_2_F3   AGTATGAGGAGGTCTGGCTGG       
PAX6_2_R3   GGCATCTTTAATGATCAGACTTGT
</code></pre>

<p>Specify the name of this file using the --primer_sequences command line parameter.</p>

<h4>
<a id="6-run-undr-rover-from-the-command-line" class="anchor" href="#6-run-undr-rover-from-the-command-line" aria-hidden="true"><span class="octicon octicon-link"></span></a>6. Run Undr Rover from the command line.</h4>

<p>An example of a command line used to run Undr Rover would be as follows:</p>

<pre><code>undr_rover.py
    --primer_coords &lt;file containing target regions&gt; 
    --primer_sequences &lt;file containing primer sequences&gt;
    --log log_file &lt;log_file_name.log&gt;
    --reference &lt;reference.fa&gt;
    --out &lt;output_file_name.vcf&gt;
    --coverdir &lt;directory to place coverage files&gt;
    sample1*.fastq sample2*.fastq
</code></pre>

<p>Sample1 and sample2 must each have 2 fastq files, corresponding to the first and second reads in a read-pair. Please be aware that only FASTQs are accepted. If you have FASTQ.gz files, they will need to be decompressed prior to using Undr Rover. </p>

<h4>
<a id="7-a-log-file-is-produced-by-undr-rover" class="anchor" href="#7-a-log-file-is-produced-by-undr-rover" aria-hidden="true"><span class="octicon octicon-link"></span></a>7. A log file is produced by Undr Rover.</h4>

<p>A log file with the specified name will be created by Undr Rover. Below is an example of the start of a log file created by Undr Rover. </p>

<pre><code>Mon, 15 Dec 2014 17:04:20 Program started.
Mon, 15 Dec 2014 17:04:20 Command line: undr_rover.py --primer_bases 15 --primer_coords k-plex_roverfile2.txt --primer_sequences k-plex_primers.txt --log log_file --id_info dbsnp.vcf.gz --reference fasta/hg19bis.fa --out variant_calls.vcf --coverdir coverage_files test_fastqs/SD0390_S70_L001_R1_001.fastq test_fastqs/SD0390_S70_L001_R2_001.fastq
Mon, 15 Dec 2014 17:04:21 Processing sample SD0390_S70_L001
Mon, 15 Dec 2014 17:04:22 Processing block chr: chr10, start: 89624214, end: 89624313
Mon, 15 Dec 2014 17:04:22 Number of read pairs in block: 0
Mon, 15 Dec 2014 17:04:22 Number of acceptable read pairs in block: 0
Mon, 15 Dec 2014 17:04:22 Number of variants found in block: 0
Mon, 15 Dec 2014 17:04:22 Processing block chr: chr10, start: 89653675, end: 89653774
Mon, 15 Dec 2014 17:04:22 Number of read pairs in block: 0
Mon, 15 Dec 2014 17:04:22 Number of acceptable read pairs in block: 0
Mon, 15 Dec 2014 17:04:22 Number of variants found in block: 0
</code></pre>

<h4>
<a id="8-example-of-a-complete-command-line-for-undr-rover" class="anchor" href="#8-example-of-a-complete-command-line-for-undr-rover" aria-hidden="true"><span class="octicon octicon-link"></span></a>8. Example of a complete command line for Undr Rover.</h4>

<pre><code>undr_rover.py   
    --primer_coords roverfile.txt 
    --primer_sequences primers.txt 
    --log log_file 
    --reference fasta/hg19bis.fa 
    --out variant_calls.vcf 
    --coverdir coverage_files 
    sample1*.fastq sample2*.fastq
</code></pre>

<p>Note that there are various optional command line arguments which can be used to alter the behaviour of Undr Rover. Further details on these can be found in the README on Undr Rover's <a href="https://github.com/rli99/undr_rover">GitHub page</a>.</p>

<p>This assumes that there are two FASTQ files each for sample1* and sample2*, which contain the first and second reads respectively for each read-pair. Roverfile.txt and primers.txt are both TSV format files containing data as described earlier. </p>

<p>The variants called will be written to the file variant_calls.vcf. This file will report called variants in accordance with the VCF v4.2 specification outlined at <a href="http://samtools.github.io/hts-specs/VCFv4.2.pdf">here</a>. Coverage files containing the number of read-pairs which mapped to each region will be output in coverage_files/sample1.coverage and coverage_files/sample2.coverage. A log file describing the actions taken by the program will be stored in log_file.</p>
        </section>

        <footer>
          Undr Rover is maintained by <a href="https://github.com/rli99">rli99</a><br>
          This page was generated by <a href="http://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
        </footer>

        
      </div>
    </div>
  </body>
</html>